keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md

```diff
@@ -0,0 +1,226 @@
+that have stricter requirements for uptime, some level of coordination is required between the two
+components to ensure that upgrading control planes does not result in running the same controller
+in more than one place while also accounting for version skew.
+
+In order to coordinate cloud-specific controllers across the kube-controller-manager and the
+cloud-controller-manager, there will be a shared resource lock between the two during leader election.
+For the cloud-controller-manager, this requires no changes since it already locks on a resource named
+`cloud-controller-manager`. However, the `kube-controller-manager` should also lock on a resource
+named `cloud-controller-manager` for the cloud specific controllers so that only 1 component in the
+cluster can run them. This requires splitting up the kube-controller-manager’s run func into
+two categories: “core” and “cloud”.
+
+<br/>
+<br/>
+<br/>
+
+![current state of leader election in kube-controller-manager](/keps/sig-cloud-provider/images/pre-kube-controller-manager-leader-election.png)
+<br/>
+**Figure 1**: current state of leader election in kube-controller-manager
+
+<br/>
+<br/>
+<br/>
+
+![proposed changes to leader election in kube-controller-manager](/keps/sig-cloud-provider/images/post-kube-controller-manager-leader-election.png)
+<br/>
+**Figure 2**: proposed changes to leader election in kube-controller-manager
+
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Splitting the Main Run Func
+
+The main run func will be split by separating the `StartControllers` function of the
+kube-controller-manager into two functions: `StartCoreControllers` and `StartCloudControllers`.
+Both functions will receive the same controller context so that the same client set and
+shared informers can be used. The call to `leaderelection.RunOrDie()` which groups the
+cloud-specific controllers will run in a go routine. The call to `leaderelection.RunOrDie()`
+which groups the core controllers will be blocking.
+
+#### Process Handling when Lock is Lost
+
+Leader election currently assumes that the elected leader must terminate if it loses
+its leader election resource lock. When the kube-controller-manager runs leader election
+against two separate resource locks (kube-controller-manager for core, and
+cloud-controller-manager for cloud-specific), losing leadership in one resource lock should
+not always warrant the current process to terminate since controllers part of the other
+leader election group may still be running. In the case of the kube-controller-manager,
+a loss in the kube-controller-manager resource lock should result in the process
+terminating (this is true today) but a loss in the cloud-controller-manager resource
+lock should only stop the set of cloud-specific controllers. This will be accomplished
+by passing in the `stopCh` to the `OnStoppedLeading` leader election call back for the
+cloud-controller-manager resource lock.
+
```

Comment:
Once a CCM has acquired a lock, how do you prevent KCM from re-acquiring the lock?

Reply:
You shouldn't have to.

Assuming a KCM to CCM HA migration 1.17 -> 1.18. When you start the lock will be held by one of the 1.17 KCMs. During the migration the lock may be held by either a 1.17 KCM or a 1.18 CCM. During the migration it is fine for the lock to be taken over by a 1.17 KCM. By the end of the migration all the 1.17 KCMs will have been taken down. At that point the lock must be held by a 1.18 CCM.